### PR TITLE
Runtime selection of multithreading vs. singlethreading

### LINKED ISSUE #57: Change threading API
See also: https://github.com/JuliaLinearAlgebra/Octavian.jl/issues/25

### PR DESCRIPTION
**Pull request by DilumAluthge**
Fixes #57 

Multi-threaded functions: (public, but not exported)
- `Gaius.mul!(C, A, B)`
- `Gaius.mul(A, B)`

Single-threaded functions: (public, but not exported)
- `Gaius.mul_single_threaded!(C, A, B)`
- `Gaius.mul_single_threaded(A, B)`

### CODE REVIEW

**Comment by MasonProtter** [comment]
I think the threaded one should be called `blocked_mul` or maybe something even shorter. The single threaded one doesn’t need a concise name imo.

**Comment by DilumAluthge** [comment]
Here are some ideas.

ideas for the multi-threaded one:
- `blocked_mul` and `blocked_mul!`
- `blockedmul` and `blockedmul!`
- `tmul` and `tmul!`

ideas for the single-threaded one:
- `blocked_mul_serial` and `blocked_mul_serial!`

**Comment by MasonProtter** [comment]
Originally, I had actaully just called it `Gaius.mul!` and `Gaius.mul` (which is why I didn't export that). I liked that API quite a bit, but I was getting confused during some testing so I switched the name. Could call it like `gmul` or something for 'Gaius mul'. 

I'm not sure what would be best.

**Comment by DilumAluthge** [comment]
@chriselrod Any thoughts on naming?

**Comment by DilumAluthge** [comment]
We could do `matmul!` for the multithreaded version, and `matmul_serial!` for the singlethreaded version.

Do any stdlibs or packages export `matmul` or `matmul!`.

**Comment by chriselrod** [comment]
`gmul!` is fine, because it is short. That's why I picked `jmul` in PaddedMatrices (`j` for Julia).
Although I wish I went with `matmul`.

I don't know of any libraries exporting `matmul` other than `Octavian`.

**Comment by DilumAluthge** [comment]
If you want, you can switch PaddedMatrices to use `matmul!`.

If you use `matmul!` in PaddedMatrices, I'll use `gmul!` in Gaius. If you stick with `jmul!` in PaddedMatrices, I'll switch Gaius to use `matmul!`.

**Comment by chriselrod** [comment]
> If you want, you can switch PaddedMatrices to use `matmul!`.
> 
> If you use `matmul!` in PaddedMatrices, I'll use `gmul!` in Gaius. If you stick with `jmul!` in PaddedMatrices, I'll switch Gaius to use `matmul!`.

There are [zero dependents](https://juliahub.com/ui/Packages/PaddedMatrices/S9CmO/0.2.3?t=2) of the latest version and [one dependent that isn't mine](https://github.com/AStupidBear/FastRecurFlux.jl) for an older version, so making the change is manageable.

I guess I'll take `matmul` and bump to 0.3.

**Comment by chriselrod** [comment]
I also like `Gaius.mul`, but `PaddedMatrices.mul` is just too long.

Honestly, I want to change the name of the library as well. It doesn't even offer a convenient way for creating padded arrays at the moment, aside from `view`, which works just as well in `Base`.

**Comment by DilumAluthge** [comment]
I like `recmul!` for Gaius, and `matmul!` for PaddedMatrices.

We should stop exporting `matmul!` from Octavian.

**Comment by chriselrod** [comment]
I do plan on exporting `matmul` and matmul!` from PaddedMatrices.

**Comment by DilumAluthge** [comment]
> Honestly, I want to change the name of the library as well. It doesn't even offer a convenient way for creating padded arrays at the moment, aside from `view`, which works just as well in `Base`.
> If I do change the name, it'd be to something like `StrideArray` or even `ArrayInterfaceArrays`, which don't sound anywhere near as good as `Gaius.mul`.

Would it be feasible to put the two different functionalities (padded arrays, and pure Julia matrix multiplication) into two separate packages?

**Comment by chriselrod** [comment]
> Would it be feasible to put the two different functionalities (padded arrays, and pure Julia matrix multiplication) into two separate packages?

Yes, we should be able to (for example) move all of the code to Octavian.

**Comment by MasonProtter** [comment]
How would you feel about just not exporting it and calling it `Gaius.mul!` @DilumAluthge?

**Comment by DilumAluthge** [comment]
> How would you feel about just not exporting it and calling it `Gaius.mul!` @DilumAluthge?

I like it. The name Gaius is short enough that I don't think it's cumbersome to type it out.

**Comment by DilumAluthge** [comment]
This is what we have now:

Multi-threaded functions: (public, but not exported)
- `Gaius.mul(A, B)`
- `Gaius.mul!(C, A, B)`

Single-threaded functions: (public, but not exported)
- `Gaius.mul_single_threaded(A, B)`
- `Gaius.mul_single_threaded!(C, A, B)`

File: README.md [hunk 1]
```diff
@@ -55,12 +72,16 @@ commonly encountered numeric `struct` types such as `Rational` and
 
 <summary>Click to expand:</summary>
 
-Gaius exports the functions `blocked_mul` and
-`blocked_mul!`. `blocked_mul` is to be used like the regular `*`
-operator between two matrices whereas `blocked_mul!` takes in three
+Gaius exports the functions `mul` and
```
Comment (chriselrod) [inline README.md:75]: "Gaius defines but does not export"?

File: README.md [hunk 2]
```diff
@@ -36,11 +36,28 @@ for learning about the implementation of linear algebra routines.
...
+
+julia> Gaius.mul_single_threaded!(C, A, B) # (single-threaded) multiply A×B and store the result in C (overwriting the contents of C)
+
+julia> Gaius.mul_single_threaded(A, B) # (single-threaded) multiply A×B and return the result
 ```
 
+Remember to start Julia with multiple threads with e.g. one of the following:
+- `julia -t auto`
+- `julia -t 4`
+- Set the `JULIA_NUM_THREADS` environment variable to `4` **before** starting Julia
```
Comment (chriselrod) [inline README.md:51]: Interesting to note that I get similar performance with 18 and 36 threads on an 18 core/36 thread system.
BLAS normally suffers with too many threads, but the cache oblivious algorithm probably makes Gaius less sensitive.

**Review by chriselrod — COMMENTED** [review]
Looks good. I like the `Threading` abstract type.

It could also be used to set recursion depth limits, e.g.

```julia
struct LimitedDepthMutliThreading <: Threading
    depth::Int
end
```
Where it starts with `depth =  8sizeof(Int)-leading_zeros(Threads.nthreads()-1)` or so, and subtracts 1 from depth on every level. If depth is 0, it calls `singlethreaded` instead.
Not saying to implement this, just that it's nice that it makes things more easily extensible.

File: README.md (hunk 1 above)
Comment (DilumAluthge) [reply README.md]: Good catch. Fixed.

**Review by chriselrod — APPROVED** [review]
Great, looks good to me.

**Review by MasonProtter — APPROVED** [review]
Looks good to me!

**Pull request merged**